sympy/solvers/ode/single.py

```diff
@@ -382,6 +381,132 @@ def inverse(self):
+                    x                y
+                    /                /
+                   |                |
+        F(x, y) =  |  P(t, y) dt +  |  Q(x0, t) dt = C1
+                   |                |
+                  /                /
+                  x0               y0
+
+    Where the first partials of `P` and `Q` exist and are continuous in a
+    simply connected region.
+
+    A note: SymPy currently has no way to represent inert substitution on an
+    expression, so the hint ``1st_exact_Integral`` will return an integral
+    with `dy`.  This is supposed to represent the function that you are
+    solving for.
+
+    Examples
+    ========
+
+    >>> from sympy import Function, dsolve, cos, sin
+    >>> from sympy.abc import x
+    >>> f = Function('f')
+    >>> dsolve(cos(f(x)) - (x*sin(f(x)) - f(x)**2)*f(x).diff(x),
+    ... f(x), hint='1st_exact')
+    Eq(x*cos(f(x)) + f(x)**3/3, C1)
+
+    References
+    ==========
+
+    - https://en.wikipedia.org/wiki/Exact_differential_equation
+    - M. Tenenbaum & H. Pollard, "Ordinary Differential Equations",
+      Dover 1963, pp. 73
+
+    # indirect doctest
+
+    """
+    hint = "1st_exact"
+    has_integral = True
+    order = [1]
+
+    def _wilds(self, f, x, order):
+        P = Wild('P', exclude=[f(x).diff(x)])
+        Q = Wild('Q', exclude=[f(x).diff(x)])
+        return P, Q
+
+    def _equation(self, fx, x, order):
+        P, Q = self.wilds()
+        return P + Q*fx.diff(x)
+
+    def _verify(self, fx) -> bool:
+        P, Q = self.wilds()
+        x = self.ode_problem.sym
+        y = Dummy('y')
+
+        m, n = self.wilds_match()
+
+        m = m.subs(fx,y)
+        n = n.subs(fx,y)
+        numerator = cancel(m.diff(y) - n.diff(x))
+        if numerator.is_zero:
+            # Is exact
+            return True
+        else:
+            # The following few conditions try to convert a non-exact
+            # differential equation into an exact one.
+            # References:
+            # 1. Differential equations with applications
+            # and historical notes - George E. Simmons
+            # 2. https://math.okstate.edu/people/binegar/2233-S99/2233-l12.pdf
+
+            factor_n = cancel(numerator/n)
+            factor_m = cancel(-numerator/m)
+            if factor_n.free_symbols == {x}:
+                # If (dP/dy - dQ/dx) / Q = f(x)
+                # then exp(integral(f(x))*equation becomes exact
+                factor = factor_n
+            elif factor_m.free_symbols == {y}:
+                # If (dP/dy - dQ/dx) / -P = f(y)
+                # then exp(integral(f(y))*equation becomes exact
+                factor = factor_m
+            else:
+                # Couldn't convert to exact
+                return False
+
+            factor = exp(Integral(factor).doit())
+            m *= factor
+            n *= factor
+            self._wilds_match[P] = m.subs(y, fx)
+            self._wilds_match[Q] = n.subs(y, fx)
```

Comment:
It looks like this substitution is undone later (in `_get_general_solution`). What is the purpose of this?

Reply:
In `_verify`, `x` and `f(x)` need to be considered independent, otherwise `free_symbols` would only return `x` and it wouldn't be possible to convert to exact. Thus the substituion is made to a dummy `y`, which is reverted in the code you highlighted.

In `_get_general_solution` we also need them being considered independent to be able to integrate, hence a second substitution of `f(x)` by `y` is performed. In the end, this substitution is also reverted in `Subs()`.

I believe this choice of two substitutions was made in the original PR to avoid using `global y` ([see original comment](https://github.com/sympy/sympy/pull/18887#issuecomment-601531785)).

Reply:
There are three places that substitute in each direction. Earlier in `_verify` we do:
```
        m = m.subs(fx,y)
        n = n.subs(fx,y)
```
(there should be space after comma)

Then here we do
```
            self._wilds_match[P] = m.subs(y, fx)
            self._wilds_match[Q] = n.subs(y, fx)
```
in order to pass these on to `_get_general_solution` but the first thing that does is:
```
        m = m.subs(fx, y)
        n = n.subs(fx, y)
```
The first of these substitutions would seem justified based on what you say. Ultimately that substitution is undone by the `Subs`. What I don't understand is the 2nd and 3rd substitutions that just seem to undo each other.

Also I think that modifying `_wilds_match` doesn't seem ideal. It would be better just to do `self.m = m` or something.

Reply:
They indeed undo each other. This redundancy is used because `_get_general_solution` doesn't know what `y` is, it only knows about `x` and `fx` (from `ode_problem`).
We pass to `_get_general_solution` as `fx` (2nd substitution), declare a new `y` and then replace it back (3rd one) for the integrals.
Removing the 2nd and 3rd substitution and just passing `m` and `n` in terms of `y` generates an error in the integrals because that `y` from `_verify` is not defined in the context of `_get_general_solution`.
Maybe this repetition could be avoided by storing `y`, but it seems weird since it's just a placeholder for `f(x)`.